### ISSUE TITLE
Slow inference when using LayerNormalization instead of BatchNormalization

### ISSUE THREAD

**Issue by pwuertz** [NVIDIA/TensorRT#3015]
## Description

I've noticed that TensorRT inference speed, in general, is significantly slower when using `LayerNormalization` instead of `BatchNormalization`. Also in particular, any gains I've seen with int8 quantization in a batch normalizing model completely vanish when switching to layer normalizations.

This is confusing because if I'm not mistaken, the difference between layer- and batch-norm is just the strategy for training the normalization factor(s) and offset(s), and both being simple multiply+add ops at inference time.

I'm trying to understand if the model import/export workflow is the problem here? I'm using Keras/Tensorflow for building and training the model, then export it as ONNX using `tf2onnx`, then read and optimize the ONNX model with TensorRT.

For a simple toy model with a `BatchNormalization` layer I see a simple ONNX conversion on the left which contains a batch-norm-layer, and sure enough, an optimized TRT engine on the right with the batch norm absorbed into one of the convolutions:
![image](https://github.com/NVIDIA/TensorRT/assets/1819283/8f816db6-8636-4cbe-a14d-aa4c469944b9)

When replacing `BatchNormalization` with `LayerNormalization`, the exported ONNX model seems to have gained a stupendous amount of complexity in order to represent the layer norm via batch norm, and it looks like that complexity carries over into the optimized engine one-to-one:
![image](https://github.com/NVIDIA/TensorRT/assets/1819283/408e0524-1100-48f2-a898-ce615ec9bdf7)

Is this the reason for the discrepancy? The complex representation of `LayerNormalization` in ONNX obscures the possibility of reducing the layer-norm-block to a fusible multiply+add op?

If so, is there a known fix for that workflow? I've tried updating `tf2onnx` to the current `1.14.0` version and exporting the model with an explicitly higher `opset` via `tf2onnx.convert.from_keras(model, opset=17)`, hoping a more expressive layer-norm-layer to appear instead of the complex sub-graph, but the ONNX graph looks the same. Is `tf2onnx` the problem here?

Are there other possible workarounds, like "freezing" normalization layers and replacing then with fixed operations for inference?

## Environment

**TensorRT Version**: 8.6.1

**NVIDIA GPU**: GeForce 4080

**NVIDIA Driver Version**: 530.30.02

**CUDA Version**: 12.1

**CUDNN Version**: 8.9

Operating System: Ubuntu 22.04

Python Version: 3.10

Tensorflow Version: 2.12

tf2onnx version: 1.14.0

## Steps To Reproduce

- Create model using Keras/Tensorflow
- Export model to ONNX using `tf2onnx`
- Import and optimize ONNX model with TensorRT
- Plot ONNX model (netron) and TensorRT Engine (trex)

**Comment by zerollzeng** [comment]
@nvpohanh Do you know? Thanks!

**Comment by nvpohanh** [comment]
In inference, BatchNorm is just a simple "Scale" op (i.e. multiply by a constant and add another constant) without needing to do any real "Reduction". In contrary, InstanceNorm and LayerNorm needs to actually do the "Reduction" (like taking the Average across some dimensions) even in the inference phase, so it is expected that InstanceNorm/LayerNorm is much slower than BatchNorm in inference.

**Comment by nvpohanh** [comment]
The `LayerNormalization` op has been added to ONNX in opset=17: https://github.com/onnx/onnx/blob/main/docs/Operators.md#LayerNormalization

So if you export the ONNX with opset >= 17, the graph should be quite clean and use the `LayerNormalization` op. However, the perf will still be similar.

**Comment by zerollzeng** [comment]
Thanks Po-Han, this is also the first answer in my mind but I'm not 100% sure. Thanks for clarify this :-)

**Comment by pwuertz** [comment]
Ah, thanks @nvpohanh for the clarification!

> So if you export the ONNX with opset >= 17, the graph should be quite clean and use the LayerNormalization op.

Right, that doesn't seem to happen with `tf2onnx.convert.from_keras(model, opset=17)` at this time, but that's probably a question for the `tf2onnx` folks. I managed to produce a toy model using `onnx_graphsurgeon`, where a `LayerNormalization` is being translated into a single TensorRt `Myelin` operation. So thats probably as clean as it gets :)

> Also in particular, any gains I've seen with int8 quantization in a batch normalizing model completely vanish when switching to layer normalizations.

That probably also answers my second question about `int8` optimization. `LayerNormalization` and its TRT implementation surely can't stay within `int8` for calculating the input mean and variance statistics. So a quantized `Conv` & `LNorm` model is going to require some `int` / `float` conversions and won't achieve the same level of `int8` speedup compared to a `Conv` & `BNorm` model.
